### ISSUE TITLE
too many namespaces

### ISSUE THREAD

**Issue by timydaley** [satijalab/seurat#548]
Hi, 
I'm having issues using Seurat with other packages with RStudio on OSX.  Trying to load other packages results in the error
maximal number of DLLs reached

It seems that Seurat loads 136 namespaces, or at least 136.  
> length(loadedNamespaces())
[1] 9
> library(Seurat)
Loading required package: ggplot2
Loading required package: cowplot

Attaching package: ‘cowplot’

The following object is masked from ‘package:ggplot2’:

    ggsave

Loading required package: Matrix
Warning message:
package ‘Matrix’ was built under R version 3.4.4 
> length(loadedNamespaces())
[1] 145

Trying to detach them results in the following error
> lapply(loadedNamespaces(), function(x) unloadNamespace(x))
Error in unloadNamespace(x) : 
  namespace ‘colorspace’ is imported by ‘munsell’ so cannot be unloaded

**Comment by leonfodoulian** [comment]
Hi Tim,

Please refer to this [stackoverflow thread](https://stackoverflow.com/questions/36974206/error-maximal-number-of-dlls-reached).

Best,
Leon

**Comment by mschilli87** [comment]
also: duplicate of https://github.com/satijalab/seurat/issues/449 (actually [not hard to find](https://github.com/satijalab/seurat/issues?utf8=%E2%9C%93&q=is%3Aissue+DLL))

**Comment by timydaley** [comment]
I saw the stack overflow thread and that did not fix it for me.

**Comment by mschilli87** [comment]
@timydaley: Would still be an R issue and not a Seurat one. So definitely the wrong place to look for help.

**Comment by mojaveazure** [comment]
@timydaley Please see #449, set the `R_MAX_NUM_DLLS` environment variable, and restart R (make sure no R/RStudio processes are running before starting R again). The `ulimit` command doesn't always work.

**Comment by timydaley** [comment]
Well, the question I would ask is, do you really need so many namespaces?  It seems to me that the package is suffering severely from bloat, which seems to be a consistent problem with so many created issues in github talking about the same problem.

**Comment by mojaveazure** [comment]
I responded in #449, stating that we cannot reduce the number packages Seurat loads. One of the great things about Seurat is how much it does in a single package. We provide a single, integrated toolkit for users to perform their analyses of single cell data. Our users don't need to wonder where that one function they need is, it's all in Seurat. To fix this DLL problem, we would need to either split Seurat into many separate packages (far less convenient for the end user and leading to greater headache for us to maintain) or to rewrite a lot of the C/C++ tools we use into R (making Seurat significantly slower than it is now). It's a problem we are aware of and would like to address; we are looking to reducing the number of packages we import, but have limited resources so cannot devote all of our time to rewriting what already exists.

**Comment by satijalab** [comment]
Hi Tim,

As Paul said, we are constantly trying to add functionality, and this does increase dependencies. Thankfully, R enables the easy modification of the R_MAX_NUM_DLLS variable.

Still, I agree that its time for some pruning. We have already started on this, and will improve in future updates. We'll try to find the right balance between being lightweight but also supporting a wide range of functionality. Thanks for the feedback.

best,
Rahul
